### PR TITLE
Improve restoring of access entities' dependencies

### PR DESCRIPTION
**Pull request by vitlibar**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve restoring of access entities' dependencies.

For example if before making a backup a role was granted to a user:
```
CREATE USER u1;
CREATE ROLE r1;
GRANT r1 TO u1;
```

then during `RESTORE` the following cases are possible:
1. User `u1` and role `r1` both don't exist. Then `RESTORE` will recreate them both and that grant too.
2. User `u1` and role `r1` both exist. Then `RESTORE` (with default settings) will skip them both and that grant too.
3. User `u1` doesn't exist and role `r1` exist. Then `RESTORE` (with default settings) will recreate user `u1`, and (after this PR) that grant too.
4. User `u1` exists and role `r1` doesn't exist. Then `RESTORE` (with default settings) will recreate role `r1`, but without granting it to existing user `u1`. This case will be improved in separate PR https://github.com/ClickHouse/ClickHouse/pull/69563

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 71d54833daf99ab79681d68f32e1fd75563f0138 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/69346/71d54833daf99ab79681d68f32e1fd75563f0138/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/69346/71d54833daf99ab79681d68f32e1fd75563f0138/style_check.html">❌ failure</a></td></tr>
<tbody>
</table>

File: src/Access/SettingsProfileElement.cpp [hunk 1]
```diff
@@ -175,6 +175,29 @@ void SettingsProfileElements::replaceDependencies(const std::unordered_map<UUID,
...
+        {
+            const UUID & profile_id = *element.parent_profile;
+            if (std::find(ids.begin(), ids.end(), profile_id) != ids.end())
+            {
+                auto f = [&](const SettingsProfileElement & x) { return x.parent_profile && *x.parent_profile == profile_id; };
+                if (std::find_if(begin(), end(), f) == end())
+                {
+                    SettingsProfileElement new_element;
+                    new_element.parent_profile = profile_id;
+                    new_elements.emplace_back(new_element);
```
Comment (qoega) [inline src/Access/SettingsProfileElement.cpp:193]:
```suggestion
                    new_elements.emplace_back(new_element);
```

File: src/Access/Role.cpp [hunk 2]
```diff
@@ -21,10 +21,19 @@ std::vector<UUID> Role::findDependencies() const
...
     settings.replaceDependencies(old_to_new_ids);
 }
 
+void Role::copyDependenciesFrom(const IAccessEntity & src, const std::vector<UUID> & ids)
+{
+    if (getType() != src.getType())
+        return;
+    const auto & src_role = typeid_cast<const Role &>(src);
+    granted_roles.copyDependenciesFrom(src_role.granted_roles, ids);
+    settings.copyDependenciesFrom(src_role.settings, ids);
```
Comment (qoega) [inline src/Access/Role.cpp:36]:
```suggestion
    settings.copyDependenciesFrom(src_role.settings, ids);
```

File: src/Access/IAccessStorage.cpp [hunk 3]
```diff
@@ -595,12 +598,50 @@ void IAccessStorage::backup(BackupEntriesCollector & backup_entries_collector, c
...
+        {
+            /// If new entities restored from backup have dependencies on other entities from backup which were not restored because they existed,
+            /// then we should correct those dependencies.
+            auto update_func = [&](const AccessEntityPtr & entity) -> AccessEntityPtr
+            {
+                auto res = entity;
+                IAccessEntity::replaceDependencies(res, new_to_existing_ids);
+                return res;
+            };
+            std::vector<UUID> ids;
```
Comment (pufit) [inline src/Access/IAccessStorage.cpp:640]:
```suggestion
            std::vector<UUID> ids;
            ids.reserve(my_entities.size());
```

File: src/Access/IAccessStorage.cpp [hunk 4]
```diff
@@ -595,12 +598,50 @@ void IAccessStorage::backup(BackupEntriesCollector & backup_entries_collector, c
...
+
+    auto entities = restorer.getAccessEntitiesToRestore();
+    if (entities.empty())
+        return;
+
+    auto create_access = restorer.getRestoreSettings().create_access;
+    bool replace_if_exists = (create_access == RestoreAccessCreationMode::kReplace);
+    bool throw_if_exists = (create_access == RestoreAccessCreationMode::kCreate);
+
+    restorer.addDataRestoreTask([this, my_entities = std::move(entities), replace_if_exists, throw_if_exists] mutable
```
Comment (pufit) [inline src/Access/IAccessStorage.cpp:617]: `my_entities` is probably not the best naming here

File: src/Access/IAccessStorage.cpp [hunk 5]
```diff
@@ -595,12 +598,50 @@ void IAccessStorage::backup(BackupEntriesCollector & backup_entries_collector, c
...
+    {
+        std::unordered_map<UUID, UUID> new_to_existing_ids;
+        for (auto & [id, entity] : my_entities)
+        {
+            UUID existing_entity_id;
+            if (!insert(id, entity, replace_if_exists, throw_if_exists, &existing_entity_id))
```
Comment (pufit) [inline src/Access/IAccessStorage.cpp:623]: It's sad that we have to change 20+ files just to have `existing_entity_id` returned from the insert, but I guess it works.

**Review by pufit — APPROVED** [review]

File: src/Access/IAccessStorage.cpp (hunk 3 above)
Comment (vitlibar) [reply src/Access/IAccessStorage.cpp:640]: I agree, but I'd rather change this and https://github.com/ClickHouse/ClickHouse/pull/69346#discussion_r1762197354 in my second PR https://github.com/ClickHouse/ClickHouse/pull/69563.
Those two corrections are not really essential, so I'd prefer not to rebuild everything.

File: src/Access/IAccessStorage.cpp (hunk 4 above)
Comment (vitlibar) [reply src/Access/IAccessStorage.cpp:617]: https://github.com/ClickHouse/ClickHouse/pull/69346#discussion_r1762785049

**Comment by vitlibar** [comment]
CI failure is wrong:
- [Bugfix validation — New test(s) failed to reproduce a bug](https://s3.amazonaws.com/clickhouse-test-reports/69346/f768717be86d0198847783f10b4c9a8f662d78d5/bugfix_validation.html) - there is a test in this PR and it definitely reproduces the bug!

**Pull request merged**